Quartz/Engine/Include/Quartz/Core/Graphics/API/GL/GLRenderDevice.hpp

```diff
@@ -0,0 +1,94 @@
+// Copyright 2019 Genten Studios
+// 
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the 
+// following conditions are met:
+// 
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the 
+// following disclaimer.
+// 
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the 
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+// 
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote 
+// products derived from this software without specific prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED 
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A 
+// PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY 
+// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR 
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH 
+// DAMAGE.
+
+#pragma once
+
+#include <Quartz/Core/Graphics/API/IRenderDevice.hpp>
+#include <Quartz/Core/Utilities/HandleAllocator.hpp>
+#include <Quartz/Core/Graphics/API/InputLayout.hpp>
+#include <Quartz/Core/Graphics/API/GL/GLVertexBuffer.hpp>
+#include <Quartz/Core/Graphics/API/GL/GLShaderPipeline.hpp>
+#include <Quartz/Core/Graphics/API/GL/GLTexture.hpp>
+
+namespace qz { namespace gfx { namespace api { namespace gl {
+	struct VertexStream {
+		VertexBufferHandle buffer;
+		int stride, offset;
+		bool active = false;
+	};
+
+	struct Uniform {
+		UniformType type;
+		GLint location;
+		ShaderPipelineHandle shader;
+		const char* name;
+		
+		union {
+			float vec2[2];
+			float vec3[3];
+			float vec4[4];
+		} userdata;
```

Comment:
Default initialize these to 0 as well, and for the union, do a default constructor like:

```cpp
Uniform()
{
  memset(this + offsetof(vec2), 0, 4 * sizeof(float));
}
```

And remember to include <cstring>.